wolfssl/src/ssl.rs

```diff
@@ -77,6 +77,88 @@ impl<IOCB: IOCallbacks> SessionConfig<IOCB> {
         }
     }
 
+    /// call `func` on `Self`
+    pub fn apply<F>(self, func: F) -> Self
+    where
+        F: FnOnce(Self) -> Self,
+    {
+        func(self)
+    }
+
+    /// When `cond` is True call `func` on `Self`
+    pub fn when<F>(self, cond: bool, func: F) -> Self
+    where
+        F: FnOnce(Self) -> Self,
+    {
+        if cond {
+            func(self)
+        } else {
+            self
+        }
+    }
+
+    /// When `cond` is True call `true_func` on `Self`, otherwise call `false_func`
+    pub fn when_or_else<T, F>(self, cond: bool, true_func: T, false_func: F) -> Self
+    where
+        T: FnOnce(Self) -> Self,
+        F: FnOnce(Self) -> Self,
+    {
+        if cond {
+            true_func(self)
+        } else {
+            false_func(self)
+        }
+    }
+
+    /// When `cond` is True call fallible `func` on `Self`
+    pub fn try_when<F>(self, cond: bool, func: F) -> Result<Self>
+    where
+        F: FnOnce(Self) -> Result<Self>,
+    {
+        if cond {
+            func(self)
+        } else {
+            Ok(self)
+        }
+    }
+
+    /// When `cond` is True call `true_func` on `Self`, otherwise call `false_func`
+    pub fn try_when_or_else<T, F>(self, cond: bool, true_func: T, false_func: F) -> Result<Self>
+    where
+        T: FnOnce(Self) -> Result<Self>,
+        F: FnOnce(Self) -> Result<Self>,
```

Comment:
Two signatures are same, we could use only one generic `T` here.

Reply:
Actually I had planned to drop the `or_else` ones since they were a bit complex and there wasn't a good usecase.

Reply:
... and I just noticed that all the methods on `ContextBuilder` are fallible and all those on `SessionConfig` aren't. So we only need the `try_` variants for the former and the non-try variants for the latter.